File: source/setup/setup_qr.c
#include "setup_qr.h"

#include <stdint.h>
#include <stdio.h>
#include <string.h>

#define HEADER_SIZE   9U
#define CHECKSUM_SIZE 4U

static uint32_t read_be32(const unsigned char *data)
{
	return ((uint32_t)data[0] << 24) | ((uint32_t)data[1] << 16) |
	       ((uint32_t)data[2] << 8) | data[3];
}

static uint32_t crc32_ieee(const unsigned char *data, size_t length)
{
	uint32_t crc = 0xFFFFFFFFU;
	for (size_t i = 0; i < length; i++) {
		crc ^= data[i];
		for (int bit = 0; bit < 8; bit++)
			crc = (crc >> 1) ^ (0xEDB88320U & (0U - (crc & 1U)));
	}
	return ~crc;
}

static bool safe_text(const unsigned char *data, size_t length)
{
	for (size_t i = 0; i < length; i++)
		if (data[i] < 0x21 || data[i] > 0x7E)
			return false;
	return true;
}

bool setup_qr_parse(const unsigned char *payload, size_t length,
	                setup_credentials *out, char *err, int errlen)
{
	if (!payload || !out || length < HEADER_SIZE + CHECKSUM_SIZE) {
		snprintf(err, errlen, "Setup QR is truncated");
		return false;
	}
	if (memcmp(payload, "SP3D", 4) != 0 || payload[4] != 1 || payload[5] != 0) {
		snprintf(err, errlen, "Unsupported setup QR");
		return false;
	}
	const size_t client_len = payload[6];
	const size_t refresh_len = ((size_t)payload[7] << 8) | payload[8];
	const size_t wanted = HEADER_SIZE + client_len + refresh_len + CHECKSUM_SIZE;
	if (!client_len || client_len > SETUP_CLIENT_ID_MAX || !refresh_len ||
	    refresh_len > SETUP_REFRESH_MAX || wanted != length) {
		snprintf(err, errlen, "Setup QR credential lengths are invalid");
		return false;
	}
	if (read_be32(payload + length - CHECKSUM_SIZE) !=
	    crc32_ieee(payload, length - CHECKSUM_SIZE)) {
		snprintf(err, errlen, "Setup QR checksum failed");
		return false;
	}
	const unsigned char *client = payload + HEADER_SIZE;
	const unsigned char *refresh = client + client_len;
	if (!safe_text(client, client_len) || !safe_text(refresh, refresh_len)) {
		snprintf(err, errlen, "Setup QR contains invalid credentials");
		return false;
	}
	memset(out, 0, sizeof *out);
	memcpy(out->client_id, client, client_len);
	memcpy(out->refresh_token, refresh, refresh_len);
	return true;
}

File: internal/setupqr/setupqr.go
package setupqr

import (
	"encoding/binary"
	"errors"
	"hash/crc32"
)

const (
	Version         = 1
	MaxClientID     = 127
	MaxRefreshToken = 255
	headerSize      = 9
	checksumSize    = 4
)

var magic = [4]byte{'S', 'P', '3', 'D'}

type Credentials struct {
	ClientID     string
	RefreshToken string
}

func Encode(credentials Credentials) ([]byte, error) {
	client := []byte(credentials.ClientID)
	refresh := []byte(credentials.RefreshToken)
	if len(client) == 0 || len(client) > MaxClientID {
		return nil, errors.New("client ID length is unsupported")
	}
	if len(refresh) == 0 || len(refresh) > MaxRefreshToken {
		return nil, errors.New("refresh token length is unsupported")
	}
	if !safe(client) || !safe(refresh) {
		return nil, errors.New("credentials contain unsupported control characters")
	}
	payload := make([]byte, headerSize+len(client)+len(refresh)+checksumSize)
	copy(payload[:4], magic[:])
	payload[4] = Version
	payload[5] = 0
	payload[6] = byte(len(client))
	binary.BigEndian.PutUint16(payload[7:9], uint16(len(refresh)))
	copy(payload[9:], client)
	copy(payload[9+len(client):], refresh)
	checksumAt := len(payload) - checksumSize
	binary.BigEndian.PutUint32(payload[checksumAt:], crc32.ChecksumIEEE(payload[:checksumAt]))
	return payload, nil
}

func Decode(payload []byte) (Credentials, error) {
	if len(payload) < headerSize+checksumSize {
		return Credentials{}, errors.New("payload is truncated")
	}
	if string(payload[:4]) != string(magic[:]) {
		return Credentials{}, errors.New("payload magic is invalid")
	}
	if payload[4] != Version || payload[5] != 0 {
		return Credentials{}, errors.New("payload version or flags are unsupported")
	}
	clientLen := int(payload[6])
	refreshLen := int(binary.BigEndian.Uint16(payload[7:9]))
	if clientLen == 0 || clientLen > MaxClientID || refreshLen == 0 || refreshLen > MaxRefreshToken {
		return Credentials{}, errors.New("credential lengths are invalid")
	}
	want := headerSize + clientLen + refreshLen + checksumSize
	if len(payload) != want {
		return Credentials{}, errors.New("payload length does not match its header")
	}
	checksumAt := want - checksumSize
	if binary.BigEndian.Uint32(payload[checksumAt:]) != crc32.ChecksumIEEE(payload[:checksumAt]) {
		return Credentials{}, errors.New("payload checksum is invalid")
	}
	client := payload[headerSize : headerSize+clientLen]
	refresh := payload[headerSize+clientLen : checksumAt]
	if !safe(client) || !safe(refresh) {
		return Credentials{}, errors.New("credentials contain unsupported control characters")
	}
	return Credentials{ClientID: string(client), RefreshToken: string(refresh)}, nil
}

func safe(value []byte) bool {
	for _, b := range value {
		if b < 0x21 || b > 0x7e {
			return false
		}
	}
	return true
}
